File: src/main/java/org/apache/spark/jruby/JRubyIteratableAdaptor.java
package org.apache.spark.jruby;

import org.jruby.Ruby;
import org.jruby.RubyEnumerator;
import org.jruby.RubyStopIteration;
import org.jruby.exceptions.RaiseException;
import org.jruby.runtime.builtin.IRubyObject;

import java.util.Iterator;
import java.util.NoSuchElementException;

import static org.apache.spark.jruby.TypeUtils.rubyToJava;

/**
 * Created by chenyh on 3/31/16.
 */
public
class JRubyIteratableAdaptor implements Iterable {
    private RubyEnumerator obj;
    private Ruby runtime;
    class JRubyIterator implements Iterator {
        IRubyObject nextObj = null;

        private boolean tryReadAhead() {
            if (nextObj != null)
                return true;
            try {
                nextObj = obj.next(runtime.getCurrentContext());
                return true;
            } catch (RaiseException e) {
                if(e.getException() instanceof RubyStopIteration) {
                    return false;
                } else {
                    throw e;
                }
            }
        }

        @Override
        public boolean hasNext() {
            return tryReadAhead();
        }

        @Override
        public Object next() {
            if(!tryReadAhead())
                throw new NoSuchElementException();
            IRubyObject n = nextObj;
            nextObj = null;
            return rubyToJava(runtime, n);
        }
    }

    @Override
    public Iterator iterator() {
        return new JRubyIterator();
    }

    public JRubyIteratableAdaptor(Ruby runtime, RubyEnumerator obj) {
        this.obj = obj;
        this.runtime = runtime;
    }
}

File: src/main/scala/org/apache/spark/jruby/JRubyIteratableAdaptor.scala
package org.apache.spark.jruby

import java.util.Iterator

import org.apache.spark.util.KnownSizeEstimation
import org.jruby.exceptions.RaiseException
import org.jruby.runtime.builtin.IRubyObject
import org.jruby.{Ruby, RubyEnumerator, RubyStopIteration}

/**
  * Created by chenyh on 4/2/16.
  */
class JRubyIteratableAdaptor[AnyRef](private val runtime: Ruby, private var obj: RubyEnumerator)
  extends java.lang.Iterable[AnyRef] with KnownSizeEstimation {

  private[jruby]
  class JRubyIterator extends java.util.Iterator[AnyRef] {
    private[jruby] var nextObj: IRubyObject = null

    private def tryReadAhead: Boolean = {
      if (nextObj != null) return true
      try {
        nextObj = obj.next(runtime.getCurrentContext)
        true
      } catch {
        case e: RaiseException =>
          if (e.getException.isInstanceOf[RubyStopIteration]) false else throw e
      }
    }

    def hasNext: Boolean = tryReadAhead

    def next: AnyRef = {
      if (!tryReadAhead) throw new NoSuchElementException
      val n: IRubyObject = nextObj
      nextObj = null
      TypeUtils.rubyToJava(runtime, n).asInstanceOf[AnyRef]
    }
  }

  override def iterator: java.util.Iterator[AnyRef] = new JRubyIterator

  override def estimatedSize: Long = 16
}
